### PR TITLE
Add support for GCC 13.1.0, bump 'testing' config to GCC 13.1.0

### PR DESCRIPTION
**Pull request by darcagn**
GCC 13.1.0-RC2 was released today with an eye to release the final version on Wednesday, April 26, 2023. 
It's time to tie up the loose ends and merge GCC 13.1.0 support!

- GCC 13.1.0 patches added, including `arm-Darwin` patch
- `config.mk.testing.sample` adjusted to GCC 13.1.0
- `README.md` adjusted to reflect changes
- GCC 12.2.0 patch moved to `historical` folder

Making this a draft PR as there are still a few things to do:
- Compile latest KOS, examples, and kos-ports against GCC 13.1.0-RC2 and test functionality
- Wait for the actual GCC 13.1.0 final tarball to be uploaded, build script won't work until then

To the last point, since the final tarball isn't uploaded yet, this can be compiled with RC2 instead like so:
```
wget http://ftp.ntua.gr/mirror/gnu/gcc/snapshots/13.1.0-RC2-20230419/gcc-13.1.0-RC2-20230419.tar.xz -O gcc-13.1.0.tar.xz
./download.sh
./unpack.sh
mv gcc-13.1.0-RC2-20230419 gcc-13.1.0
cd ./gcc-13.1.0 && ./contrib/download_prerequisites && cd ..
```
Ignore the `./unpack.sh: line 78: cd: ./gcc-13.1.0: No such file or directory` error while unpacking.

Lastly, Windows support hasn't been tested yet -- if you get a moment, can you test and make patches if necessary for Windows support, @sizious ?

### CODE REVIEW

**Comment by darcagn** [comment]
I've confirmed through my testing that GCC 13.1.0 RC2 is building under mingw32 without any extra platform-specific patching necessary. 

GCC 13.1.0 RC2 with KOS patches tested building on Windows (mingw32), macOS, Linux
GCC 13.1.0 RC2 raw toolchain tested building on macOS, Linux
KOS, kos-ports, and all examples are building without error

Just have to go through and test running all of the examples now.

**Comment by gyrovorbis** [comment]
Just going to let people know that I've finally taken the time to test the new GCC13 prerelease toolchain...

...and everything seems to work perfectly. No problems building it, KOS, KOS-ports, or examples. All examples I've tested (maybe half a dozen+) are working perfectly.

ONE THING I WANT TO NOTE HERE:
There was an ICE (internal compiler error) in GCC12 that caused it to hang and blow up your RAM usage until your OS terminated it when you tried to build any version of SDL I could get my hands on with it. This prevented me from ever building KOS or kos-utils with `-O3` or `-O4` globally...

Well, it's been resolved! So the only compiler issue I even knew of in 12 is already fixed in 13!

**Comment by sizious** [comment]
So I guess 12 couldn’t be a ‘stable’ candidate?

**Comment by darcagn** [comment]
> So I guess 12 couldn’t be a ‘stable’ candidate?

It's only been available in KOS since early February, and as such, personally I don't think it's been tested enough by the overall community to qualify as stable. I think we should move to gcc 13.1.0 for testing and keep 9.3.0 for stable. When gcc 14.x comes out in a year, 13.x will have been tested for a year and could be moved to stable and 14.x can move to testing. What's your opinion, @ljsebald ?

**Comment by darcagn** [comment]
I've marked this PR as ready for review. 

Before the actual merge can happen, the following needs to be done:
- [ ] Verify KOS, kos-ports, and all examples can build
- [ ] Verify all examples run with no regressions from GCC 12.2.0
- [ ] GCC 13.1.0 tarball is uploaded and the build script can retrieve it

I'd like to have this PR already reviewed so that once the tarball is uploaded on Wednesday, April 26, I can can immediately merge.

**Comment by ljsebald** [comment]
> > So I guess 12 couldn’t be a ‘stable’ candidate?
> 
> It's only been available in KOS since early February, and as such, personally I don't think it's been tested enough by the overall community to qualify as stable. I think we should move to gcc 13.1.0 for testing and keep 9.3.0 for stable. When gcc 14.x comes out in a year, 13.x will have been tested for a year and could be moved to stable and 14.x can move to testing. What's your opinion, @ljsebald ?

Seems reasonable.

File: utils/dc-chain/patches/gcc-13.1.0-kos.diff [hunk 1]
```diff
@@ -0,0 +1,1552 @@
...
+diff --color -ruN gcc-13.1.0/libgcc/config/sh/crt1.S gcc-13.1.0-kos/libgcc/config/sh/crt1.S
+--- gcc-13.1.0/libgcc/config/sh/crt1.S	2023-04-06 09:48:33.091053980 -0500
++++ gcc-13.1.0-kos/libgcc/config/sh/crt1.S	2023-04-06 09:47:57.481945701 -0500
+@@ -1,724 +1,225 @@
+-/* Copyright (C) 2000-2023 Free Software Foundation, Inc.
+-   This file was pretty much copied from newlib.
++! KallistiOS ##version##
++!
++! startup.s
++! (c)2000-2001 Dan Potter
```
Comment (ljsebald) [inline utils/dc-chain/patches/gcc-13.1.0-kos.diff:37]: This line should get updated.

Comment (darcagn) [reply utils/dc-chain/patches/gcc-13.1.0-kos.diff]: Got it! Nice eye.

**Comment by gyrovorbis** [comment]
I've just verified source-level debugging with GDB works great on it and got my Qt Creator setup going for GCC-13.

**Comment by sizious** [comment]
Tested to build that on my MinGW-w64 x86 GCC 9.x for DreamSDK and it works as well so I will be able to add GCC 13 in my package too. 😉

**Comment by gyrovorbis** [comment]
So I finally got my massive libGimbal unit test suites compiling for the new toolchain. I used them last time to validate GCC12, because they flex my entire libGimbal framework, all of the C11 standard library, and have over 800 unit tests for everything it entails, so it's a great test for toolchain verification. It also does profiling for each test. Because they flex so much, they're more representative of a complex app, and I wanted to make sure they worked before blessing this toolchain. 

If you want to view the tests, they're here: https://github.com/gyrovorbis/libgimbal/tree/master/test/source. Here's the example output for a couple of the suites so you can see what they look like:
```
* ********* Starting TestSuite [GblSortTestSuite] *********
* [ INIT      ]: GblSortTestSuite
* [ RUN       ]: GblSortTestSuite::bubbleSort
* [      PASS ]: GblSortTestSuite::bubbleSort (139.000 ms)
* [ RUN       ]: GblSortTestSuite::selectionSort
* [      PASS ]: GblSortTestSuite::selectionSort (33.000 ms)
* [ RUN       ]: GblSortTestSuite::insertionSort
* [      PASS ]: GblSortTestSuite::insertionSort (52.000 ms)
* [ RUN       ]: GblSortTestSuite::shellSort
* [      PASS ]: GblSortTestSuite::shellSort (19.000 ms)
* [ RUN       ]: GblSortTestSuite::mergeSort
* [      PASS ]: GblSortTestSuite::mergeSort (14.000 ms)
* [ RUN       ]: GblSortTestSuite::quickSort
* [      PASS ]: GblSortTestSuite::quickSort (12.000 ms)
* [ FINAL     ]: GblSortTestSuite
* Totals: 6 passed, 0 failed, 0 skipped, 269.000ms, 0/0 leaked (0/0 bytes)
* ********* Finished TestSuite [GblSortTestSuite] *********
* ********* Starting TestSuite [GblHashTestSuite] *********
* [ INIT      ]: GblHashTestSuite
* [ RUN       ]: GblHashTestSuite::fnv1
* [      PASS ]: GblHashTestSuite::fnv1 (1.000 ms)
* [ RUN       ]: GblHashTestSuite::murmur
* [      PASS ]: GblHashTestSuite::murmur (2.000 ms)
* [ RUN       ]: GblHashTestSuite::superFast
* [      PASS ]: GblHashTestSuite::superFast (1.000 ms)
* [ RUN       ]: GblHashTestSuite::pearson
* [      PASS ]: GblHashTestSuite::pearson (1.000 ms)
* [ RUN       ]: GblHashTestSuite::jenkins
* [      PASS ]: GblHashTestSuite::jenkins (1.000 ms)
* [ RUN       ]: GblHashTestSuite::sip
* [      PASS ]: GblHashTestSuite::sip (2.000 ms)
* [ RUN       ]: GblHashTestSuite::crc
* [      PASS ]: GblHashTestSuite::crc (5.000 ms)
* [ FINAL     ]: GblHashTestSuite
* Totals: 7 passed, 0 failed, 0 skipped, 13.000ms, 0/0 leaked (0/0 bytes)
* ********* Finished TestSuite [GblHashTestSuite] *********
```

Thank GOD everything works beautifully on GCC13. The results are in! Here's the final output of the test suite.

GCC 12.2.0:
```
* ********* Finished TestSuite [GblThreadTestSuite] *********
* [GblTestScenario] Ending Scenario: [libGimbalTests]
        * Runtime Statistics
                * Total Time (ms)     :              874.000
                * Max Allocations     :                  110
                * Max Allocation Size :                 6144
                * Max Allocated Bytes :                16376
                * Remaining Allocs    :                   52
                * Remaining Bytes     :                14652
                * Seed                :           3663974473
        * Test Suite Totals
                * Total               :                   45
                * Passed              :                   45
                * Skipped             :                    0
                * Failed              :                    0
        * Test Case Totals
                * Total               :                  876
                * Passed              :                  875
                * Skipped             :                    1
                * Failed              :                    0
* ********************* [   PASS   ] *********************

```
GCC 13.3.1
```
* ********* Finished TestSuite [GblThreadTestSuite] *********
* [GblTestScenario] Ending Scenario: [libGimbalTests]
        * Runtime Statistics
                * Total Time (ms)     :              738.000
                * Max Allocations     :                  110
                * Max Allocation Size :                 6144
                * Max Allocated Bytes :                16376
                * Remaining Allocs    :                   52
                * Remaining Bytes     :                14652
                * Seed                :           3663974306
        * Test Suite Totals
                * Total               :                   45
                * Passed              :                   45
                * Skipped             :                    0
                * Failed              :                    0
        * Test Case Totals
                * Total               :                  876
                * Passed              :                  875
                * Skipped             :                    1
                * Failed              :                    0
* ********************* [   PASS   ] *********************
```

At least for `-O2`, GCC13 seems to consistently win by about 150ms off of the total execution time, which is a pretty big deal. I'd have to look into the individual test cases to see what things in particular 13 is faster with, and I could do some more timing at other optimization levels, but I just wanted to share these cursory findings, and say that I now give this toolchain my blessing and will be using it for my own stuff.

**Review by gyrovorbis — DISMISSED** [review]
Just make sure the URL and download.sh work fine when it's released, but other than that, I approve.

**Comment by darcagn** [comment]
> So I finally got my massive libGimbal unit test suites compiling for the new toolchain. I used them last time to validate GCC12, because they flex my entire libGimbal framework, all of the C11 standard library, and have over 800 unit tests for everything it entails, so it's a great test for toolchain verification. It also does profiling for each test. Because they flex so much, they're more representative of a complex app, and I wanted to make sure they worked before blessing this toolchain.

Thanks for this! Everything falls in line with my testing. I've done a compiler test showdown pitting 4.7.4 vs. 9.3.0 vs. 12.2.0 vs. 13.1.0 and things are really looking good for 13.1.0: https://dcemulation.org/phpBB/viewtopic.php?t=106068

**Review by gyrovorbis — APPROVED** [review]

**Review by ljsebald — APPROVED** [review]

**Pull request merged**